packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -967,3 +967,85 @@ abstract class PopupRoute<T> extends ModalRoute<T> {
     super.didChangeNext(nextRoute);
   }
 }
+
+/// A [Navigator] observer that notifies [RouteAware]s of changes to the
+/// state of their [Route].
+///
+/// To make a [StatefulWidget] aware of its current [Route] state, implement
+/// [RouteAware] in its [State] and subscribe it to the [RouteObserver]:
+///
+/// ```dart
+/// class RouteAwareWidgetState extends State<RouteAwareWidget> with RouteAware {
+///   @override
+///   void didChangeDependencies() {
+///     super.didChangeDependencies();
+///     routeObserver.subscribe(this, ModalRoute.of(context));
+///   }
+///
+///   @override
+///   void dispose() {
+///     routeObserver.unsubscribe(this);
+///     super.dispose();
+///   }
+///
+///   @override
+///   void didPush() {
+///     // Do something
+///   }
+///
+///   @override
+///   void didPopNext() {
+///     // Do something
+///   }
+///
+/// }
+/// ```
+class RouteObserver<T extends Route<dynamic>> extends NavigatorObserver {
+  final Map<T, RouteAware> _listeners = <T, RouteAware>{};
+
+  void subscribe(RouteAware routeAware, T route) {
+    assert(routeAware != null);
+    assert(route != null);
+    if (!_listeners.containsKey(route)) {
+      routeAware.didPush();
```

Comment:
why do we call didPush when you subscribe?

Reply:
This code was originally written by @weisong0 and then migrated into Flutter.

It looks like the intention is for `subscribe()` to be called in `dependenciesChanged()`, which is the earliest point that a `State` has become active and can safely call `ModalRoute.of(context)`.

Example code here: https://github.com/flutter/plugins/blob/master/packages/firebase_analytics/example/lib/tabs_page.dart#L35

It's looks a bit odd to me that the corresponding call to `didPop()` occurs in `didPop()` instead of `unsubscribe()`. @weisong0, do you think it would make sense to move the `didPop()` call to `unsubscribe()`?

Reply:
`unsubscribe()` is intended to be called in `dispose()`. However, I am not sure when `dispose()` occurs after the associated route is popped. If `dispose()` is called every time the route is popped, then I think it makes sense to move it to `unsubscribe()`.

Reply:
I filed https://github.com/flutter/flutter/issues/13919 to track this